### PR TITLE
Fix for invocation when calling npm under sudo

### LINKED ISSUE #4290: NPM Resource does not report on local package installations properly
The NPM resource allows a path to be set to find the node_modules in a project directory. The command that it uses works on the command line.

cd <PROJECT_DIR> && npm ls --json <PACKAGE_NAME>

```json
{
  "dependencies": {
    "sprintf-js": {
      "version": "1.1.2",
      "from": "sprintf-js@latest",
      "resolved": "https://registry.npmjs.org/sprintf-js/-/sprintf-js-1.1.2.tgz"
    }
  }
}
```

But when run in InSpec the test states that the package is not installed:

```ruby
describe npm('azure-storage', path: '/usr/local/camsa/statsd/azure-queue') do
   it { should be_installed }
end
```

```
  Npm Package azure-storage
     [FAIL]  should be installed
     expected that `Npm Package azure-storage` is installed
```

### PR DESCRIPTION
**Pull request by clintoncwolfe**
Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

## Description

When calling `npm`, which happens when using the `npm` resource, it was possible for the invocation to fail due to `"sudo: cd: command not found\n"` when a path was provided to the `npm` resource.

The fix wraps the call in a `sh -c` wrapper.

Additionally, on windows, the resource never worked properly when a custom path was provided, since it was using `&&` to cd then check for the module. This PR now uses `;` to separate the cd command on Windows, which PowerShell seems to tolerate.

I've manually tested this on Windows; I have a very brittle kitchen / Vagrant setup that does everything but install the actual NPM modules, which had to be done manually.

Screen Shot 2019-11-07 at 17 29 54

## Related Issue

Fixes #4290 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New content (non-breaking change)
- [x] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

**Review by zenspider — APPROVED** [review]

File: lib/inspec/resources/npm.rb [hunk 1]
```diff
@@ -32,7 +32,7 @@ def info
         npm = "npm -g"
       end
 
-      cmd = inspec.command("#{npm} ls --json #{@package_name}")
+      cmd = inspec.command("sh -c '#{npm} ls --json #{@package_name}'")
```
Comment (miah) [inline lib/inspec/resources/npm.rb]: We should probably remove 'windows' as a supported platform.

**Review by miah — APPROVED** [review]

File: lib/inspec/resources/npm.rb (hunk 1 above)
Comment (russellseymour) [reply lib/inspec/resources/npm.rb]: @miah I do not think that removing support for Windows is the way to go here.

Is it not be possible to detect the platform that Inspec is running on and choose if `sh -c` should be used for Linux or `cmd /c` for Windows?

Comment (clintoncwolfe) [reply lib/inspec/resources/npm.rb]: Thank you, on further reflection what I want to express is "there has never been correct support for windows if you use a custom path", but by removing it here we'd remove the npm resource for Windows all users, custom path or no. 

I agree, a custom Windows-specific invocation is needed.

**Comment by miah** [comment]
related: https://github.com/inspec/inspec/issues/3749

File: lib/inspec/resources/npm.rb (hunk 1 above)
Comment (clintoncwolfe) [reply lib/inspec/resources/npm.rb]: This is now in place and ready for review.

**Review by miah — APPROVED** [review]

File: lib/inspec/resources/npm.rb [hunk 2]
```diff
@@ -27,12 +27,20 @@ def info
...
       else
-        npm = "npm -g"
+        invocation = "npm -g"
       end
 
-      cmd = inspec.command("#{npm} ls --json #{@package_name}")
+      invocation = "#{invocation} ls --json #{@package_name}"
+
+      # If on unix, wrap in sh -c to protect against sudo
+      unless inspec.os.platform?("windows")
```
Comment (zenspider) [inline lib/inspec/resources/npm.rb:39]: I generally eschew `unless` if there is a positive way to describe what I want:

```
# If on unix, wrap in sh -c to protect against sudo
unless inspec.os.platform?("windows")
```

vs

```
# Wrap in sh -c to protect against sudo
if inspec.os.platform?("unix")
```

Seems clearer to me.

File: lib/inspec/resources/npm.rb [hunk 3]
```diff
@@ -27,12 +27,20 @@ def info
...
+        invocation = "sh -c '#{invocation}'"
+      end
+
+      cmd = inspec.command(invocation)
```
Comment (zenspider) [inline lib/inspec/resources/npm.rb:43]: Is this something that `command` (or train?) should be doing across the board? If on unix and sudo then wrap w/ `sh`?

**Pull request merged**